### ISSUE TITLE
Are n>1 queries supported with the server?

### ISSUE THREAD

**Issue by arnavsinghvi11** [huggingface/text-generation-inference#736]
### System Info

OS Version:
Distributor ID: Ubuntu
Description: Ubuntu 20.04.3 LTS
Release: 20.04
Codename: focal

8 A-100 GPUS

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

I'm running docker run --gpus all --shm-size 1g -p 8070:80 -v $volume:/data -e HUGGING_FACE_HUB_TOKEN=<my token> ghcr.io/huggingface/text-generation-inference:0.9.3 --model-id $model --num-shard $num_shard --max-input-length 4000 --max-total-tokens 4096

I'm running this on the latest commit: https://github.com/huggingface/text-generation-inference/commit/1da642bd0e6de28ef499f17cd226264f3ccdc824

### Expected behavior

I was wondering if n > 1 queries are supported with the server currently based on how I'm passing in the kwargs to the server. 

For instance, this is part of my current generate function that takes in kwargs that would include n queries being set along with other parameters. So if n > 1, is this supported with the text generation inference server currently with batched queries to the server?

`payload = {
        "inputs": prompt,
        "parameters": {
            **kwargs,
            "do_sample": False,
            "temperature": 0.7,
            "top_p": 0.9,
            "stop":["\n", "\n\n"]
            }
        }`
        
`response = requests.post(self.url + "/generate", json=payload, headers=self.headers)
`

**Comment by Narsil** [comment]
> n > 1 queries are supported with the server currently based on how I'm passing in the kwargs to the server

Do you mean sending a list of prompts ? Like `["Hello", " Good morning"]` and expect to receive 2 different conversations ?
This is currently not supported and not planned.
Sending 2 different queries should still end up batching anyway.

Adding support would mean complexifying a lot the streaming support (currently we can drop requests when 1 query is finished).
Also during the stream we would need to give a lot more information about for which query a token is linked.

What is your actual underlying problem without it ? There might be other ways around it.

**Comment by okhat** [comment]
Not quite. We want to send a single prompt and generate more than one completion, @Narsil

**Comment by Narsil** [comment]
We have `best_of: n ` which will generate `n` replies and choose the best (logits wise).

We do not have something like `num_return_sequences: 5` or `num_beams: 5` at the moment.

**Comment by okhat** [comment]
Thanks for the fast reply. That's a pretty serious limitation for things like doing chain of thought and self-consistency, @Narsil

**Comment by Narsil** [comment]
Hmm.. seems I forgot about it  `num_return_sequences` does exist (it's just `best_of`).

```
curl 127.0.0.1:8080/generate     -X POST     -d '{"inputs":"parameters":{"max_new_tokens":30, "best_of": 2, "top_k": 20, "details": true}}'  -H 'Content-Type: application/json'
```

Will work.

**Comment by Narsil** [comment]
(By doing that you cannot use the streaming option though obviously)

**Comment by okhat** [comment]
Awesome! Thanks @Narsil !

So... best_of=5 is actually going to return 5 things, if details=True? This sounds great for us, but maybe docs need to clarify this.

**Comment by Narsil** [comment]
Indeed.

And yes we are in the process to clear up docs and make them better ! 
In the meantime: https://github.com/huggingface/text-generation-inference/blob/main/clients/python/text_generation/client.py#L61

**Comment by okhat** [comment]
@Narsil we're seeing

```
Failed to parse JSON response: {"error":"Input validation error: `best_of` must be > 0 and <= 2. Given: 5","error_type":"validation"}
```

Only 1 or 2 is accepted :sweat:

**Comment by robertvacareanu** [comment]
@okhat This is because the default for `max_best_of` is 2. 

Passing `--max-best-of` with the `docker run` command will fix this issue:
```
docker run --gpus all --shm-size 1g -p 8080:80 -v $volume:/data ghcr.io/huggingface/text-generation-inference:1.0.0 --model-id $model --max-best-of 100
```

Then, if you request `best_of=5`, you will have 4 completions inside the `['details']['best_of_sequence']` which will be different from the one inside `['generated_text']`.

**Comment by okhat** [comment]
Thanks a lot Robert!

**Comment by LopezGG** [comment]
> > n > 1 queries are supported with the server currently based on how I'm passing in the kwargs to the server
> 
> Do you mean sending a list of prompts ? Like `["Hello", " Good morning"]` and expect to receive 2 different conversations ? This is currently not supported and not planned. Sending 2 different queries should still end up batching anyway.
> 
> Adding support would mean complexifying a lot the streaming support (currently we can drop requests when 1 query is finished). Also during the stream we would need to give a lot more information about for which query a token is linked.
> 
> What is your actual underlying problem without it ? There might be other ways around it.

How does batching work internally. If I send in 1200 requests through AsyncClient, will it automatically get batched internally
